### PR TITLE
Enforce code coverage

### PR DESCRIPTION
**Pull request by ccaominh**
### Description

Add an automated way of checking if new code has adequate unit tests, since merging code coverage reports and check coverage thresholds via coveralls or codecov is unreliable.

The following minimum unit test code coverage is now enforced on new code:
- 80% functions
- 65% branch
- 65% line

Branch and line coverage thresholds are slightly lower for now as they are harder to achieve.

After the code coverage check looks reliable, the thresholds can be increased later if needed.

<hr>

This PR has:
- [x] been self-reviewed.
- [x] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [x] been manually tested: https://github.com/ccaominh/druid/pull/1

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -136,20 +135,51 @@ jobs:
...
+        - for f in ${project_files}; do echo $f; done  # for debugging
+        # Check diff code coverage for the maven projects being tested (retry install in case of network error)
+        - >
+          if [ -n "${project_files}" ]; then
+          travis_retry npm install @connectis/diff-test-coverage@1.5.3
+          && git diff origin/${TRAVIS_BRANCH}...HEAD -- ${project_files}
+          | node_modules/.bin/diff-test-coverage
+          --coverage "**/target/site/jacoco/jacoco.xml"
+          --type jacoco
+          --line-coverage 65
```
Comment (maytasm) [inline .travis.yml:176]: Maybe making these variables at the top with the other export

File: benchmarks/pom.xml [hunk 2]
```diff
@@ -220,6 +220,13 @@
           <skip>true</skip>
         </configuration>
       </plugin>
+      <plugin>
+        <groupId>org.jacoco</groupId>
+        <artifactId>jacoco-maven-plugin</artifactId>
+        <configuration>
+          <skip>true</skip> <!-- ignore non-production code -->
```
Comment (maytasm) [inline benchmarks/pom.xml:227]: Should the integration-tests be skipped too?

File: .travis.yml [hunk 3]
```diff
@@ -136,20 +135,51 @@ jobs:
...
+        - >
+          if [[ "${MAVEN_PROJECTS}" = \!* ]]; then
+          regex="${MAVEN_PROJECTS:1}";
+          regex="^${regex//,\!/\\|^}";
+          project_files="$(echo "${all_files}" | grep -v "${regex}" || [[ $? == 1 ]])";
+          else
+          regex="^${MAVEN_PROJECTS//,/\\|^}";
+          project_files="$(echo "${all_files}" | grep "${regex}" || [[ $? == 1 ]])";
+          fi
+        - for f in ${project_files}; do echo $f; done  # for debugging
```
[the remark this answers is not in the document]
Comment (ccaominh) [reply .travis.yml:167]: Travis already echos the command to the log: https://travis-ci.org/github/ccaominh/druid/jobs/686402944#L5741

Does that accomplish what you had in mind?

File: .travis.yml [hunk 4]
```diff
@@ -136,20 +135,51 @@ jobs:
...
         - sh -c "dmesg | egrep -i '(oom|out of memory|kill process|killed).*' -C 1 || exit 0"
         - free -m
-      after_success: &upload_java_unit_test_coverage
         - ${MVN} -pl ${MAVEN_PROJECTS} jacoco:report
+        # Add merge target branch to determine diff (see https://github.com/travis-ci/travis-ci/issues/6069)
+        - echo "TRAVIS_BRANCH=${TRAVIS_BRANCH}"  # for debugging
+        - git remote set-branches --add origin ${TRAVIS_BRANCH} && git fetch
+        # Determine the modified files that match the maven projects being tested
+        - all_files="$(git diff --name-only origin/${TRAVIS_BRANCH}...HEAD | grep "\.java$" || [[ $? == 1 ]])"
+        - for f in ${all_files}; do echo $f; done  # for debugging
```
[the remark this answers is not in the document]
Comment (ccaominh) [reply .travis.yml:157]: Travis already echos the command to the log: https://travis-ci.org/github/ccaominh/druid/jobs/686402944#L5731

Does that accomplish what you had in mind?

File: .travis.yml (hunk 1 above)
Comment (ccaominh) [reply .travis.yml:176]: The `before_script` section that exports a variable is differs for the various jobs. Keeping the threshold here is nice for having defined in one place.

File: benchmarks/pom.xml (hunk 2 above)
Comment (ccaominh) [reply benchmarks/pom.xml:227]: Jacoco coverage isn't collected for the integration tests.

**Pull request merged**